src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -35,28 +37,55 @@ namespace Microsoft.CodeAnalysis;
     /// </summary>
     private readonly ImmutableSortedDictionary<DocumentId, TState> _map;
 
-    private TextDocumentStates(ImmutableList<DocumentId> ids, ImmutableSortedDictionary<DocumentId, TState> map)
+    private readonly FrozenDictionary<string, OneOrMany<DocumentId>> _filePathToDocumentIds;
```

Comment:
moved to a mapping stored in the TextDocumentStates object.  This mapping is *mcuh* easier to manage.

Reply:
note: TextDocumentStates is currently a readonly struct.  We might want to consider it a mutable-class.  That way we can even defer creation of this till needed.